File: RPC/2020:06:13 Junio/editorial/rainbowstrings/submissions/timelimitexceeded/rainbowstrings_hamerly_tle.py
#!/usr/bin/env python3

# slow DP

import string
import sys

# DP, probably should be TLE due to the 2^26 factor
# runtime: O(n + 2^k * k^2), where k <= 26 is the number of unique characters

MOD = 998244353

s = sys.stdin.readline().strip()
counts = [s.count(c) for c in string.ascii_lowercase if s.count(c)]

lengths = [(1, 0)] # count, start ndx
total = 1

for length in range(1, len(counts) + 1):
    new_lengths = []
    for count, start in lengths:
        for i, c in enumerate(counts[start:], start):
            a = (c * count) % MOD
            new_lengths.append((a, i + 1))
            total += a
    lengths = new_lengths
    total %= MOD

print(total)

File: RPC/2020:06:13 Junio/editorial/rainbowstrings/submissions/timelimitexceeded/rainbowstrings_hamerly_tle.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// DP, probably should be TLE due to the 2^26 factor
// runtime: O(n + 2^k * k^2), where k <= 26 is the number of unique characters

size_t const MOD = 998244353;

int main() {
    std::string s;
    getline(std::cin, s);
    std::vector<int> counts(26, 0);
    for (auto c : s)
        counts[c - 'a']++;
    std::sort(counts.begin(), counts.end(), std::greater<int>());
    while (counts.size() and (counts.back() == 0))
        counts.pop_back();

    std::vector<std::pair<size_t, size_t>> lengths = { {1, 0} };
    std::vector<std::pair<size_t, size_t>> new_lengths;
    size_t total = 1;

    for (size_t length = 1; length <= counts.size(); ++length) {
        new_lengths.clear();
        for (auto x : lengths) {
            for (size_t i = x.second; i < counts.size(); ++i) {
                size_t a = (x.first * counts[i]) % MOD;
                new_lengths.push_back({a, i + 1});
                total = (total + a) % MOD;
            }
        }
        lengths.swap(new_lengths);
    }

    std::cout << total << std::endl;

    return 0;
}

